Sources/UberAuth/Authorize/AuthenticationSession.swift

```diff
@@ -0,0 +1,84 @@
+//
+//  Copyright © Uber Technologies, Inc. All rights reserved.
+//
+
+
+import AuthenticationServices
+import Foundation
+
+/// @mockable
+protocol AuthenticationSessioning {
+    init(anchor: ASPresentationAnchor,
+         callbackURLScheme: String,
+         url: URL,
+         completion: @escaping AuthCompletion)
+    
+    func start()
+}
+
+final class AuthenticationSession: AuthenticationSessioning {
+    
+    private let authSession: ASWebAuthenticationSession
+    
+    private let presentationContextProvider: ASWebAuthenticationPresentationContextProviding?
+    
+    init(anchor: ASPresentationAnchor = ASPresentationAnchor(),
+         callbackURLScheme: String, 
+         url: URL,
+         completion: @escaping AuthCompletion) {
+        self.presentationContextProvider = AuthPresentationContextProvider(anchor: anchor)
+        self.authSession = ASWebAuthenticationSession(
+            url: url,
+            callbackURLScheme: callbackURLScheme,
+            completionHandler: { url, error in
+                switch (url, error) {
+                case (_, .some(let error)):
+                    completion(.failure(UberAuthError(error: error)))
+                case (.none, _):
+                    completion(.failure(UberAuthError.invalidAuthCode))
+                case (.some(let url), _):
+                    guard let code = Self.parse(url: url) else {
+                        completion(.failure(UberAuthError.invalidAuthCode))
+                        return
+                    }
+                    completion(.success(.init(authorizationCode: code)))
+                }
+            }
+        )
+        self.authSession.presentationContextProvider = presentationContextProvider
+    }
+    
+    func start() {
+        if #available(iOS 13.4, *) {
+            guard authSession.canStart else {
+                return
```

Comment:
Might be useful to log or inform the client about whether session can be or has been started

Reply:
As discussed, will update in a later PR